cpp/include/cuspatial/detail/cartesian_product_iterator.cuh

```diff
@@ -0,0 +1,227 @@
+ *     : v0   v1 : v2   v3   v4 : v5 :
+ *     +         :              :    +
+ *     : w0   w1 : w2   w3   w4 : w5 :
+ *     +---------+--------------+----+
+ *  A1 : x0   x1 : x2   x3   x4 : x5 :
+ *     +---------+--------------+----+
+ *  A2 : y0   y1 : y2   y3   y4 : y5 :
+ *     +         :              :    +
+ *     : z0   z1 : z2   z3   z4 : z5 :
+ *     +---------+--------------+----+
+ *
+ * Order of iteration:
+ *
+ *  u0, v0, w0, u1, v1, w1, u2, v2, w2, u3, v3, w3, u4, v4, w4, u5, v5, w5, x0, x1, x2, x3, x4, x5,
+ *  y0, z0, y1, z1, y2, z2, y3, z3, y4, z4, y5, z5
+ *
+ * i.e:
+ *
+ *  A0 * B0 | u0 v0 w0 u1 v1 w1
+ *  A0 * B1 | u2 v2 w2 u3 v3 w3 u4 v4 w4
+ *  A0 * B2 | u5 v5 w5
+ *  A1 * B0 | x0 x1
+ *  A1 * B1 | x2 x3 x4
+ *  A1 * B2 | x5
+ *  A1 * B0 | y0 z0 y1 z1
+ *  A2 * B1 | y2 z2 y3 z3 y4 z4
+ *  A2 * B2 | y5 z5
+ *
+ * @tparam OffsetIteratorA
+ * @tparam OffsetIteratorB
+ * @param num_elements_a  number of elements from all groups in A
+ * @param num_elements_b  number of elements from all groups in B
+ * @param num_groups_a    number of groups in A
+ * @param num_groups_b    number of groups in B
+ * @param group_offsets_a offsets for each group in A, plus an "end" offset
+ * @param group_offsets_b offsets for each group in B, plus an "end" offset
+ * @return auto           Thrust iterator of `cartesian_product_group_index`s. device-only.
+ *
+ * @note Can be used in conjunction with `make_transform_iterator` to create a single kernel which
+ * simultaneously computes distance metrics for multiple spaces in O(n^2) time using a single call
+ * to `reduce_by_key`.
+ */
+template <typename OffsetIteratorA, typename OffsetIteratorB>
+auto make_grouped_cartesian_product_iterator(int32_t const num_elements_a,
+                                             int32_t const num_elements_b,
+                                             int32_t const num_groups_a,
+                                             int32_t const num_groups_b,
+                                             OffsetIteratorA const group_offsets_a,
+                                             OffsetIteratorB const group_offsets_b)
+{
+  auto count_iter = thrust::make_counting_iterator(0);
+
+  using SizeFunctorLHS = detail::size_from_offsets_functor<OffsetIteratorA>;
+  using SizeFunctorRHS = detail::size_from_offsets_functor<OffsetIteratorB>;
+
+  auto group_sizes_a = thrust::make_transform_iterator(
+    count_iter, SizeFunctorLHS{num_groups_a, num_elements_a, group_offsets_a});
+  auto group_sizes_b = thrust::make_transform_iterator(
+    count_iter, SizeFunctorRHS{num_groups_b, num_elements_b, group_offsets_b});
+  auto group_lookup_a = thrust::make_transform_iterator(
+    count_iter, group_lookup_functor<OffsetIteratorA>{group_offsets_a, num_groups_a});
+  auto group_lookup_b = thrust::make_transform_iterator(
+    count_iter, group_lookup_functor<OffsetIteratorB>{group_offsets_b, num_groups_b});
+
+  using SizeIteratorLHS   = decltype(group_sizes_a);
```

Comment:
Why do you need these `using` definitions if they are only used once and they are brief anyway? Just put the `decltype` calls inline...